pkg/controller/jaeger/jaeger_controller.go

```diff
@@ -352,3 +356,75 @@ func (r ReconcileJaeger) getSecretsForNamespace(secrets []corev1.Secret, namespa
 	}
 	return secretsForNamespace
 }
+
+// SyncOnJaegerChanges sync deployments with sidecars when a jaeger CR changes
+func (r *ReconcileJaeger) SyncOnJaegerChanges(object client.Object) []reconcile.Request {
+	deps := []appsv1.Deployment{}
+	nssupdate := []corev1.Namespace{}
+	nss := map[string]corev1.Namespace{} // namespace cache
+
+	jaeger, ok := object.(*v1.Jaeger)
+	if !ok {
+		return []reconcile.Request{}
+	}
+
+	deployments := appsv1.DeploymentList{}
+	err := r.rClient.List(context.Background(), &deployments)
+	if err != nil {
+		return []reconcile.Request{}
+	}
+
+	for _, dep := range deployments.Items {
+		// if there's an assigned instance to this deployment, and it's not the one that triggered the current event,
+		// we don't need to trigger a reconciliation for it
+		if val, ok := dep.Labels[inject.Label]; ok && val != jaeger.Name {
+			continue
+		}
+
+		// if the deployment has the sidecar annotation, trigger a reconciliation
+		if ok := increaseRevision(dep.Annotations); ok {
+			deps = append(deps, dep)
+			continue
+		}
+
+		// if we don't have the namespace in the cache yet, retrieve it
+		var ns corev1.Namespace
+		if ns, ok = nss[dep.Namespace]; !ok {
+			err := r.rClient.Get(context.Background(), types.NamespacedName{Name: dep.Namespace}, &ns)
+			if err != nil {
+				continue
+			}
+			nss[ns.Name] = ns
+		}
+
+		// if the namespace has the sidecar annotation, trigger a reconciliation
+		if ok := increaseRevision(ns.Annotations); ok {
+			nssupdate = append(nssupdate, ns)
+			continue
+		}
+	}
+	for _, dep := range deps {
+		if err := r.client.Update(context.Background(), &dep); err != nil {
```

Comment:
I only have one concern with this approach, it could be the case that the deployment change its state between the time you get it and the time you updated it (for instance from the initial state to RUNNING) . In that case you will get an error like this:

`ERRO[1123] failed to update deployment with sidecar      error="Operation cannot be fulfilled on deployments.apps \"simple-prod-query\": the object has been modified; please apply your changes to the latest version and try again" jaeger=simple-prod jaeger-namespace=monitoring name=simple-prod-query namespace=monitoring`

Because the version is not the same, I think in a normal reconciliation process you can handle this by returning an error , which eventually will trigger a new reconciliation loop. but I'm not sure how are we handling this case with this approach.